File: src/yuvdiff/diff.py
"""Pixel-level absolute diff between two YUV frames."""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from yuvdiff.formats import PixelFormat
from yuvdiff.parser import YUVFrame


@dataclass
class DiffResult:
    """Output of DiffEngine.diff()."""

    diff_y: np.ndarray  # absolute diff of Y plane, shape (H, W)
    diff_u: np.ndarray  # shape (H/v, W/h) — native chroma resolution
    diff_v: np.ndarray
    mask: np.ndarray    # bool, shape (H, W) — True where any channel > threshold
    diff_pixel_count: int  # Strictly d > 0 pixels count
    total_pixel_count: int  # H * W
    diff_ratio: float = 0.0
    diff_gt_2t: int = 0
    diff_gt_2t_ratio: float = 0.0
    diff_gt_t: int = 0
    diff_gt_t_ratio: float = 0.0
    diff_gt_half_t: int = 0
    diff_gt_half_t_ratio: float = 0.0
    diff_mean: float = 0.0
    diff_median: float = 0.0
    diff_max: int = 0
    diff_min: int = 0


class DiffEngine:
    """Per-channel absolute diff + threshold-based mask.

    The mask combines Y and upsampled U/V with a logical OR. Threshold
    is uniform across channels in v1.
    """

    def __init__(self, threshold: int = 4):
        if threshold < 0:
            raise ValueError(f"threshold must be >= 0, got {threshold}")
        self.threshold = threshold

    def diff(self, a: YUVFrame, b: YUVFrame) -> DiffResult:
        if (a.width, a.height) != (b.width, b.height):
            raise ValueError(
                f"Frame size mismatch: A=({a.width}x{a.height}) "
                f"vs B=({b.width}x{b.height})"
            )
        if a.pixel_format != b.pixel_format:
            raise ValueError(
                f"Pixel format mismatch: A={a.pixel_format} vs B={b.pixel_format}"
            )

        diff_y = np.abs(a.y.astype(np.int32) - b.y.astype(np.int32)).astype(np.int32)
        diff_u = np.abs(a.u.astype(np.int32) - b.u.astype(np.int32)).astype(np.int32)
        diff_v = np.abs(a.v.astype(np.int32) - b.v.astype(np.int32)).astype(np.int32)

        h, w = a.y.shape
        diff_u_full = self._upsample_plane(diff_u, a.pixel_format, h, w)
        diff_v_full = self._upsample_plane(diff_v, a.pixel_format, h, w)
        diff_pixel = np.maximum(diff_y, np.maximum(diff_u_full, diff_v_full))

        total = h * w
        non_zero_mask = diff_pixel > 0
        diff_pixel_count = int(non_zero_mask.sum())
        diff_ratio = diff_pixel_count / total if total > 0 else 0.0

        t = self.threshold
        t_half = t // 2
        t_2t = t * 2

        diff_gt_2t = int((diff_pixel > t_2t).sum())
        diff_gt_2t_ratio = diff_gt_2t / total if total > 0 else 0.0

        diff_gt_t = int((diff_pixel > t).sum())
        diff_gt_t_ratio = diff_gt_t / total if total > 0 else 0.0

        diff_gt_half_t = int((diff_pixel > t_half).sum())
        diff_gt_half_t_ratio = diff_gt_half_t / total if total > 0 else 0.0

        mask = diff_pixel > t

        if diff_pixel_count > 0:
            non_zero_diffs = diff_pixel[non_zero_mask]
            diff_mean = float(np.mean(non_zero_diffs))
            diff_median = float(np.median(non_zero_diffs))
            diff_max = int(np.max(non_zero_diffs))
            diff_min = int(np.min(non_zero_diffs))
        else:
            diff_mean = 0.0
            diff_median = 0.0
            diff_max = 0
            diff_min = 0

        return DiffResult(
            diff_y=diff_y,
            diff_u=diff_u,
            diff_v=diff_v,
            mask=mask,
            diff_pixel_count=diff_pixel_count,
            total_pixel_count=total,
            diff_ratio=diff_ratio,
            diff_gt_2t=diff_gt_2t,
            diff_gt_2t_ratio=diff_gt_2t_ratio,
            diff_gt_t=diff_gt_t,
            diff_gt_t_ratio=diff_gt_t_ratio,
            diff_gt_half_t=diff_gt_half_t,
            diff_gt_half_t_ratio=diff_gt_half_t_ratio,
            diff_mean=diff_mean,
            diff_median=diff_median,
            diff_max=diff_max,
            diff_min=diff_min,
        )

    @staticmethod
    def _upsample_plane(
        plane: np.ndarray, fmt: PixelFormat, h: int, w: int
    ) -> np.ndarray:
        """Nearest-neighbor upsample a chroma plane to Y's resolution."""
        if fmt == PixelFormat.YUV444P:
            return plane
        if fmt == PixelFormat.YUV422P:
            return np.kron(plane, np.ones((1, 2), dtype=plane.dtype))
        return np.kron(plane, np.ones((2, 2), dtype=plane.dtype))

File: cpp/src/diff.cpp
#include "yuvdiff/diff.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace yuvdiff {

DiffEngine::DiffEngine(int threshold) : threshold_(threshold) {
    if (threshold < 0) {
        throw std::invalid_argument("threshold must be >= 0, got " + std::to_string(threshold));
    }
}

DiffResult DiffEngine::diff(const YUVFrame& a, const YUVFrame& b) const {
    if (a.width != b.width || a.height != b.height) {
        throw std::invalid_argument(
            "Frame size mismatch: A=(" + std::to_string(a.width) + "x" + std::to_string(a.height) +
            ") vs B=(" + std::to_string(b.width) + "x" + std::to_string(b.height) + ")"
        );
    }

    int w = a.width;
    int h = a.height;
    auto [hf_a, vf_a] = chroma_subsampling(a.pixel_format);
    auto [hf_b, vf_b] = chroma_subsampling(b.pixel_format);

    int hf = std::min(hf_a, hf_b);
    int vf = std::min(vf_a, vf_b);
    int cw = w / hf;
    int ch = h / vf;

    int scale_a = (a.bit_depth == 8 && b.bit_depth == 10) ? 4 : 1;
    int scale_b = (b.bit_depth == 8 && a.bit_depth == 10) ? 4 : 1;

    DiffResult result;
    result.diff_y.resize(static_cast<size_t>(w) * h);
    result.diff_u.resize(static_cast<size_t>(cw) * ch);
    result.diff_v.resize(static_cast<size_t>(cw) * ch);
    result.diff_pixel.resize(static_cast<size_t>(w) * h);
    result.mask.resize(static_cast<size_t>(w) * h, 0);
    result.total_pixel_count = static_cast<int64_t>(w) * h;
    result.diff_pixel_count = 0;

    // Diff Y plane
    for (int r = 0; r < h; ++r) {
        for (int c = 0; c < w; ++c) {
            int idx = r * w + c;
            int32_t val_a = static_cast<int32_t>(a.get_y(r, c)) * scale_a;
            int32_t val_b = static_cast<int32_t>(b.get_y(r, c)) * scale_b;
            result.diff_y[idx] = std::abs(val_a - val_b);
        }
    }

    // Diff U and V planes
    for (int r = 0; r < ch; ++r) {
        int vr = r * vf;
        for (int c = 0; c < cw; ++c) {
            int vc = c * hf;
            int idx = r * cw + c;
            int32_t u_a = static_cast<int32_t>(a.get_u(vr / vf_a, vc / hf_a)) * scale_a;
            int32_t u_b = static_cast<int32_t>(b.get_u(vr / vf_b, vc / hf_b)) * scale_b;
            result.diff_u[idx] = std::abs(u_a - u_b);

            int32_t v_a = static_cast<int32_t>(a.get_v(vr / vf_a, vc / hf_a)) * scale_a;
            int32_t v_b = static_cast<int32_t>(b.get_v(vr / vf_b, vc / hf_b)) * scale_b;
            result.diff_v[idx] = std::abs(v_a - v_b);
        }
    }

    // Generate combined threshold mask & combined pixel difference stats (calculated for d > 0)
    int64_t non_zero_cnt = 0;
    int64_t sum_diff_non_zero = 0;
    int32_t min_diff_non_zero = 1000000;
    int32_t max_diff_non_zero = 0;
    std::vector<int> hist_non_zero(1024, 0);

    result.threshold = threshold_;
    result.threshold_half = threshold_ / 2;
    result.threshold_2t = threshold_ * 2;

    int64_t diff_gt_t_cnt = 0;
    int64_t diff_gt_half_t_cnt = 0;
    int64_t diff_gt_2t_cnt = 0;

    for (int r = 0; r < h; ++r) {
        for (int c = 0; c < w; ++c) {
            int y_idx = r * w + c;
            int32_t dy = result.diff_y[y_idx];
            int32_t du = std::abs(
                static_cast<int32_t>(a.get_u(r / vf_a, c / hf_a)) * scale_a -
                static_cast<int32_t>(b.get_u(r / vf_b, c / hf_b)) * scale_b
            );
            int32_t dv = std::abs(
                static_cast<int32_t>(a.get_v(r / vf_a, c / hf_a)) * scale_a -
                static_cast<int32_t>(b.get_v(r / vf_b, c / hf_b)) * scale_b
            );

            int32_t d_pixel = std::max({dy, du, dv});
            result.diff_pixel[y_idx] = d_pixel;

            if (d_pixel > 0) {
                non_zero_cnt++;
                sum_diff_non_zero += d_pixel;
                if (d_pixel < min_diff_non_zero) min_diff_non_zero = d_pixel;
                if (d_pixel > max_diff_non_zero) max_diff_non_zero = d_pixel;
                if (d_pixel < 1024) {
                    hist_non_zero[d_pixel]++;
                } else {
                    hist_non_zero[1023]++;
                }

                if (d_pixel > result.threshold_half) {
                    diff_gt_half_t_cnt++;
                }
                if (d_pixel > result.threshold) {
                    diff_gt_t_cnt++;
                }
                if (d_pixel > result.threshold_2t) {
                    diff_gt_2t_cnt++;
                }
            }

            if (d_pixel > threshold_) {
                result.mask[y_idx] = 1;
            }
        }
    }

    result.diff_pixel_count = non_zero_cnt;
    result.diff_ratio = (result.total_pixel_count > 0)
        ? (static_cast<double>(result.diff_pixel_count) / result.total_pixel_count)
        : 0.0;

    result.diff_gt_half_t = diff_gt_half_t_cnt;
    result.diff_gt_half_t_ratio = (result.total_pixel_count > 0)
        ? (static_cast<double>(result.diff_gt_half_t) / result.total_pixel_count)
        : 0.0;

    result.diff_gt_t = diff_gt_t_cnt;
    result.diff_gt_t_ratio = (result.total_pixel_count > 0)
        ? (static_cast<double>(result.diff_gt_t) / result.total_pixel_count)
        : 0.0;

    result.diff_gt_2t = diff_gt_2t_cnt;
    result.diff_gt_2t_ratio = (result.total_pixel_count > 0)
        ? (static_cast<double>(result.diff_gt_2t) / result.total_pixel_count)
        : 0.0;

    if (non_zero_cnt > 0) {
        result.diff_min = min_diff_non_zero;
        result.diff_max = max_diff_non_zero;
        result.diff_mean = static_cast<double>(sum_diff_non_zero) / non_zero_cnt;

        // Fast and exact median from non-zero histogram
        if (non_zero_cnt % 2 == 1) {
            size_t target = non_zero_cnt / 2 + 1;
            size_t accum = 0;
            for (int i = 1; i < 1024; ++i) {
                accum += hist_non_zero[i];
                if (accum >= target) {
                    result.diff_median = static_cast<double>(i);
                    break;
                }
            }
        } else {
            size_t target1 = non_zero_cnt / 2;
            size_t target2 = non_zero_cnt / 2 + 1;
            size_t accum = 0;
            int val1 = -1;
            int val2 = -1;
            for (int i = 1; i < 1024; ++i) {
                accum += hist_non_zero[i];
                if (val1 == -1 && accum >= target1) {
                    val1 = i;
                }
                if (val2 == -1 && accum >= target2) {
                    val2 = i;
                    break;
                }
            }
            result.diff_median = (val1 + val2) / 2.0;
        }
    } else {
        result.diff_min = 0;
        result.diff_max = 0;
        result.diff_mean = 0.0;
        result.diff_median = 0.0;
    }

    return result;
}

} // namespace yuvdiff
